packages/loader/container-loader/src/container.ts

```diff
@@ -494,6 +505,59 @@ export class Container
 
 	private readonly mc: MonitoringContext;
 
+	/**
+	 * The current generation of the Loader layer. This is used to ensure compatibility between the Loader and
+	 * other layers.
+	 */
+	private readonly generation = currentLoaderGeneration;
+	/**
+	 * A list of features that the Runtime layer is required to support to be compatible with this Loader.
+	 */
+	private readonly requiredFeaturesFromRuntime: string[] = [];
+	/**
+	 * A list of features supported by the Loader layer. This is exposed to the Runtime layer which uses
+	 * it to determine if the Loader is compatible with the Runtime.
+	 */
+	private readonly supportedFeaturesForRuntime: Map<string, unknown> = new Map([
+		/**
+		 * This version of the loader accepts `referenceSequenceNumber`, provided by the container runtime,
+		 * as a parameter to the `submitBatchFn` and `submitSummaryFn` functions.
+		 * This is then used to set the reference sequence numbers of the submitted ops in the DeltaManager.
+		 */
+		["referenceSequenceNumbers", true],
+	]);
+	private validateRuntimeCompatibility(
+		runtimeSupportedFeatures?: ReadonlyMap<string, unknown>,
+		runtimeVersion?: string,
+	): void {
+		if (runtimeSupportedFeatures === undefined) {
```

Comment:
This surprises me... if the runtime doesn't even list any supported features, as soon as we add a required feature above, we should fail compat.  So this seems like not a useful early return, since we'll have to remove it at that point (and `checkLayerCompat` should work with an empty `runtimeSupportedFeatures` map)

Reply:
Good point. I changed the validation to support old layers until we add a required feature. Please take a look.